### PR TITLE
Feature/bk sdk without web view

### PR DESCRIPTION
**Pull request by mouli-bluekai**
Changes to have the option to make calls directly to tags server. 

Refactored the code.
Have refactored the sample app to use Android Preferences instead of Database. 
Added check in resume(), so that it does not create a web view every time (if WebViews are getting used)
Removed dependency on android support v4 jar, which was just getting bundled for displaying alert dialog for sandbox url with dev mode on when using WebView.

CR: @antonbk @burkaygur

### CODE REVIEW

File: sdk/src/com/bluekai/sdk/BlueKai.java [hunk 1]
```diff
@@ -511,21 +627,159 @@ private void sendData(Map<String, String> paramsMap) {
...
+						onDataPosted(!response.isError(), "Data posted successfully. Response: " + response.getResponseBody(), existingData, paramsList);
+					}
+				}
+			});
+			BKResponse response;
+			try {
+				response = webServiceTask.execute(request).get();
+				returnedMessage = response.getResponseBody();
+			} catch (InterruptedException e) {
+				e.printStackTrace();
```
Comment (antonbk) [inline sdk/src/com/bluekai/sdk/BlueKai.java]: hmm, I am sure we do not want to print the stack.

Comment (antonbk) [inline sdk/src/com/bluekai/sdk/BlueKai.java]: we can use `Logger`

Comment (mouli-bluekai) [inline sdk/src/com/bluekai/sdk/BlueKai.java]: Changed that.

File: sdk/src/com/bluekai/sdk/bktag/CoreTagUtil.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+			result = email.trim().toLowerCase();
+			String[] tmp = result.split("@");
+			String pre = tmp[0];
+			if (pre.indexOf("+") > -1) {
+				pre = pre.substring(0, pre.indexOf("+"));
+			}
+			result = pre + "@" + tmp[1];
+		} catch (Exception e) {
+
+		}
```
Comment (antonbk) [inline sdk/src/com/bluekai/sdk/bktag/CoreTagUtil.java]: We had one-liner in your hashing examples:
```java
return email.trim().toLowerCase().replaceFirst("\\+[^@]*@", "@");
```

File: sdk/src/com/bluekai/sdk/helpers/BKAdvertisingIdClient.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	public static AdInfo getAdvertisingIdInfo(final Context context) throws Exception {
+		if (Looper.myLooper() == Looper.getMainLooper()) throw new IllegalStateException("Cannot be called from the main thread");
+
+		try {
+			PackageManager pm = context.getPackageManager();
+			pm.getPackageInfo("com.android.vending", 0);
+		} catch (Exception e) {
+			throw e;
```
Comment (antonbk) [inline sdk/src/com/bluekai/sdk/helpers/BKAdvertisingIdClient.java]: Interesting... Why do we have try-catch than?

Comment (mouli-bluekai) [inline sdk/src/com/bluekai/sdk/helpers/BKAdvertisingIdClient.java]: Oh..I was logging it and then decided against it. Forgot to remove the try catch.

File: sdk/src/com/bluekai/sdk/bktag/CoreTagUtil.java (hunk 2 above)
Comment (mouli-bluekai) [inline sdk/src/com/bluekai/sdk/bktag/CoreTagUtil.java]: Changed that.

File: sample/src/com/bluekai/sampleapp/BlueKaiTab.java [hunk 4]
```diff
@@ -127,33 +125,43 @@ public void onClick(View v) {
...
 
 	@Override
 	public void onDataPosted(boolean success, String message) {
 		Log.d("BlueKaiSampleApp", String.valueOf(success) + " :: " + message);
+		Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
+	}
+
+	@Override
+	public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
+		if (key.equals("enableDevMode")) {
```
Comment (antonbk) [inline sample/src/com/bluekai/sampleapp/BlueKaiTab.java]: Reverse version is null-safe:
```java
"enableDevMode".equals(key)
```

File: sdk/src/com/bluekai/sdk/BlueKai.java [hunk 5]
```diff
@@ -486,10 +598,14 @@ private void checkForExistingData() {
...
-			thread.start();
+		if (useWebView) {
+			if (settings.isAllowDataPosting()) {
+				SendData sendData = new SendData(paramsList, handler, true);
+				Thread thread = new Thread(sendData);
+				thread.start();
+			}
+		} else {
+			sendDataWithoutWebView(paramsList, true);
 		}
```
Comment (antonbk) [inline sdk/src/com/bluekai/sdk/BlueKai.java]: whole if-else should be wrapped in `settings.isAllowDataPosting()`, I do not see that sendDataWithoutWebView honors the opt-in flag on its own

File: sdk/src/com/bluekai/sdk/BlueKai.java [hunk 6]
```diff
@@ -511,21 +627,157 @@ private void sendData(Map<String, String> paramsMap) {
...
+
+				@Override
+				public void afterReceivingResponse(BKResponse response) {
+					Logger.debug(TAG, "Received response: " + response.getResponseBody());
+					if (response.isError()) {
+						onDataPosted(!response.isError(), "Problem posting data", existingData, paramsList);
+					} else {
+						onDataPosted(!response.isError(), "Data posted successfully. Response: " + response.getResponseBody(), existingData, paramsList);
+					}
+				}
```
Comment (antonbk) [inline sdk/src/com/bluekai/sdk/BlueKai.java]: I see some code duplication between `sendDataWithoutWebView` and `sendDataWithoutWebViewSync `.
`BKWebServiceListener ` definition is identical.

File: sdk/src/com/bluekai/sdk/BlueKai.java (hunk 5 above)
Comment (mouli-bluekai) [inline sdk/src/com/bluekai/sdk/BlueKai.java]: Changed it. Actually, that method (sendExistingData) itself won't be called if allow data posting it not true.

File: sdk/src/com/bluekai/sdk/BlueKai.java (hunk 6 above)
Comment (mouli-bluekai) [inline sdk/src/com/bluekai/sdk/BlueKai.java]: Refactored

File: sample/src/com/bluekai/sampleapp/BlueKaiTab.java [hunk 7]
```diff
@@ -110,15 +101,22 @@ public void onClick(View v) {
...
-						Toast.makeText(context, "Value is empty. Please enter a value", Toast.LENGTH_LONG).show();
+						Toast.makeText(BlueKaiTab.this, "Value is empty. Please enter a value", Toast.LENGTH_LONG).show();
 					} else {
-						bk.put(key, value);
+						if (sync) {
+							String response = bk.putSync(key, value);
+							Toast.makeText(BlueKaiTab.this, response, Toast.LENGTH_SHORT).show();
+						} else {
+							bk.put(key, value);
+						}
```
Comment (antonbk) [inline sample/src/com/bluekai/sampleapp/BlueKaiTab.java]: Can continue else if:
```java
else if (sync) {
	String response = bk.putSync(key, value);
	Toast.makeText(BlueKaiTab.this, response, Toast.LENGTH_SHORT).show();
} else {
	bk.put(key, value);
}
```

Comment (mouli-bluekai) [inline sample/src/com/bluekai/sampleapp/BlueKaiTab.java]: I think separating sounds more logical, as the validations are grouped and the final sending of data is in the last else block.

Comment (antonbk) [inline sample/src/com/bluekai/sampleapp/BlueKaiTab.java]: ok, if this helps to navigate quicker.

File: sdk/src/com/bluekai/sdk/BlueKai.java [hunk 8]
```diff
@@ -716,27 +990,22 @@ public void onViewLoaded(boolean success, boolean existingData, ParamsList param
...
+		GetAdvertisingIdTask task = new GetAdvertisingIdTask(context, new BKAdvertisingIdListener() {
 
-		if ((androidId == null) || (androidId.equals("")) || (androidId.equals("0") || (androidId.startsWith("00000")))) {
-			androidId = getPsudeoIMEI();
-		}
-		return androidId;
-	}
+			@Override
+			public void onReceivedAdvertisingId(AdInfo adInfo) {
+				advertisingIdRetrieved = true;
```
Comment (antonbk) [inline sdk/src/com/bluekai/sdk/BlueKai.java]: Should we move `advertisingIdRetrieved ` down under `if (adInfo != null)`?

Comment (mouli-bluekai) [inline sdk/src/com/bluekai/sdk/BlueKai.java]: Yes, moved.

File: sdk/src/com/bluekai/sdk/helpers/BKAdvertisingIdClient.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+		intent.setPackage("com.google.android.gms");
+		if (context.getApplicationContext().bindService(intent, connection, Context.BIND_AUTO_CREATE)) {
+
+			try {
+
+				AdvertisingInterface adInterface = new AdvertisingInterface(connection.getBinder());
+				AdInfo adInfo = new AdInfo(adInterface.getId(), adInterface.isLimitAdTrackingEnabled(true));
+				return adInfo;
+			} catch (Exception exception) {
+				throw exception;
```
Comment (antonbk) [inline sdk/src/com/bluekai/sdk/helpers/BKAdvertisingIdClient.java]: if we just `throw exeption`, we don't need to `catch`, 
```java
try {
...
} finally {
...
}
```

Comment (mouli-bluekai) [inline sdk/src/com/bluekai/sdk/helpers/BKAdvertisingIdClient.java]: Yes, removed it.

File: sdk/src/com/bluekai/sdk/task/BKWebServiceRequestTask.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+		HttpResponse response;
+		BKResponse bkResponse = new BKResponse();
+
+		String url = request.getUrl();
+
+		try {
+			switch (request.getType()) {
+				case POST: {
+					HttpPost post = new HttpPost(url);
+					post.setHeader("Content-type", request.getContentType());
```
Comment (antonbk) [inline sdk/src/com/bluekai/sdk/task/BKWebServiceRequestTask.java]: Even though HTTP header names are case-insensitive, most commonly used version is camel-case with dashes: `Content-Type`

**Comment by antonbk** [comment]
Overall looks good to me, :shipit:

**Pull request merged**